docs/articles/dockerfile_best-practices.md

```diff
@@ -146,16 +146,33 @@ with backslashes.
 Probably the most common use-case for `RUN` is an application of `apt-get`.
 When using `apt-get`, here are a few things to keep in mind:
 
-* Don’t do `RUN apt-get update` on a single line. This will cause
-caching issues if the referenced archive gets updated, which will make your
-subsequent `apt-get install` fail without comment.
-
 * Avoid `RUN apt-get upgrade` or `dist-upgrade`, since many of the “essential”
```

Comment:
Thanks @charleswhchan  for the contribution. Most of the `RUN` material is taken up with a discussion of `apt-get` which is the real topic here.  It would be better to give a heading to `apt-get` and raise its importance. This goes beyond the scope of your original change but if you are up to expanding a bit your help and technical review is invaluable.

---> https://gist.github.com/moxiegirl/06f680798666910c9749

### RUN

[Dockerfile reference for the RUN instruction](https://docs.docker.com/reference/builder/#run)

As always, to make your `Dockerfile` more readable, understandable, and
maintainable, put long or complex `RUN` statements on multiple lines separated
with backslashes.


### apt-get 

Probably the most common use-case for `RUN` is an application of `apt-get`. The
`RUN apt-get` command, because it installs packages, has several gotchas to look
out for.

You should avoid `RUN apt-get upgrade` or `dist-upgrade`, as many of the
“essential” packages from the base images won't upgrade inside an unprivileged
container. If a base package is out-of-date, you should contact its maintainers.
If you know there’s a particular package, `foo`, that needs to be updated, use
`apt-get install -y foo` to update automatically.

Always combine  `RUN apt-get update` with `apt-get install` in the same `RUN`
statement, for example:

        RUN apt-get update && apt-get install -y \
            package-bar \
            package-baz \
            package-foo


Using `apt-get update` alone causes caching issues and subsequent `apt-get
install` instructions fail. For example, say you have a Dockerfile:

        FROM ubuntu:14.04
        RUN apt-get update
        RUN apt-get install -y curl

After building the image, all layers are in the Docker cache. Suppose you later
modify `apt-get install` by adding extra package:

        FROM ubuntu:14.04
        RUN apt-get update
        RUN apt-get install -y curl nginx

Docker sees the initial and modified instructions as identical and reuses the
cache from previous steps. As a result the `apt-get update` is *NOT* executed
because the build uses the cached version. Because the `apt-get update` is not
run, your build can potentially get an outdated version of the `curl` and `nginx`
packages.

Using  `RUN apt-get update && apt-get install -y` ensures your Dockerfile
installs the latest package versions without no further coding or manual
intervention. This technique is known as "cache busting." You can also achieve
cache-busting by specifying a package version. This is known as version pinning,
for example:

        RUN apt-get update && apt-get install -y \
            package-bar \
            package-baz \
            package-foo=1.3.*

Version pinning forces the build to retrieve a particular version regardless of
what’s in the cache. This technique can also reduce failures due to unanticipated changes
in required packages.

Below is a well-formed `RUN` instruction that demonstrates the all the `apt-get`
recommendations. 

    RUN apt-get update && apt-get install -y \
        aufs-tools \
        automake \
        build-essential \
        curl \
        dpkg-sig \
        libcap-dev \
        libsqlite3-dev \
        lxc=1.0* \
        mercurial \
        reprepro \
        ruby1.9.1 \
        ruby1.9.1-dev \
        s3cmd=1.1.0*

The `s3cmd` instructions specifies a version `1.1.0*`. If the image previously
used an older version, specifying the new one causes a cache bust of `apt-get
update` and ensure the installation of the new version. Writing on package on
each line can also prevent mistakes in package duplication.